libvirt/domain.go

```diff
@@ -0,0 +1,171 @@
+package libvirt
+
+import (
+	"encoding/xml"
+	"errors"
+
+	"fmt"
+
+	"github.com/sandlbn/libvirt-go"
+)
+
+var memInfo = struct {
+	m map[string]int32
+}{m: map[string]int32{
+	"swap_in":        0,
+	"swap_out":       1,
+	"major_fault":    2,
+	"min_fault":      3,
+	"unused":         4,
+	"available":      5,
+	"actual_balloon": 6,
+	"rss":            7,
+	"nr":             8,
+}}
+
+// GetNetworkStatistics return GoBlockStatistics
+func GetNetworkStatistics(domain libvirt.VirDomain, paths ...string) (map[string]libvirt.VirDomainInterfaceStats, error) {
+	netStat := make(map[string]libvirt.VirDomainInterfaceStats)
+	var err error
+	if len(paths) == 0 {
+		paths, err = GetDomainInterfaces(domain)
+		if err != nil {
+			return netStat, err
+		}
+	}
+	for _, path := range paths {
+		stat, err := domain.InterfaceStats(path)
+		if err != nil {
+			return netStat, err
+		}
+		netStat[path] = stat
+	}
+	return netStat, nil
+}
+
+// GetBlockStatistics return GoBlockStatistics
+func GetBlockStatistics(domain libvirt.VirDomain, paths ...string) (map[string]libvirt.VirDomainBlockStats, error) {
+	blockStat := make(map[string]libvirt.VirDomainBlockStats)
+	var err error
+	if len(paths) == 0 {
+		paths, err = GetDomainDisks(domain)
+		if err != nil {
+			return blockStat, err
+		}
+	}
+	for _, path := range paths {
+		stat, err := domain.BlockStats(path)
+		if err != nil {
+			return blockStat, err
+		}
+		blockStat[path] = stat
+	}
+	return blockStat, nil
+}
+
+//GetDomainInterfaces returns slice of available network devices
+func GetDomainInterfaces(domain libvirt.VirDomain) ([]string, error) {
+	var interfaces []string
+	domXML, err := getDomainXML(domain)
+	if err != nil {
+		return interfaces, err
+	}
+	if domXML.Devices.Interface == nil || domXML.Devices.Interface.Target == nil {
```

Comment:
It seems that you should check if `domXML.Devices` is not `nil` before this line.